source/common/http/match_delegate/config.h

```diff
@@ -29,15 +32,21 @@ class DelegatingStreamFilter : public Envoy::Http::StreamFilter {
     void evaluateMatchTree(MatchDataUpdateFunc data_update_func);
     bool skipFilter() const { return skip_filter_; }
     void onStreamInfo(const StreamInfo::StreamInfo& stream_info) {
-      if (has_match_tree_ && matching_data_ == nullptr) {
+      if (matching_data_ == nullptr) {
         matching_data_ = std::make_shared<Envoy::Http::Matching::HttpMatchingDataImpl>(stream_info);
       }
     }
+
+    void setMatchTree(Matcher::MatchTreeSharedPtr<Envoy::Http::HttpMatchingData> match_tree) {
+      match_tree_ = std::move(match_tree);
```

Comment:
nit: let's add a comment saying  something like `the matcher from per route config takes precedent position and will override the matcher from filter config(if exists)`